test/jdk/com/sun/jdi/SuspendAfterDeath.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8287847
+ * @summary Test suspending a thread after it has terminated.
+ * @enablePreview
+ * @requires vm.continuations
+ * @run build TestScaffold VMConnection TargetListener TargetAdapter
+ * @run compile SuspendAfterDeath.java
+ * @run main/othervm SuspendAfterDeath
+ */
+import com.sun.jdi.*;
+import com.sun.jdi.event.*;
+import com.sun.jdi.request.*;
+import java.util.*;
+
+class SuspendAfterDeathTarg {
+    static final String THREAD_NAME = "duke";
+
+    // breakpoint here
+    static void done() {
+    }
+
+    public static void main(String[] args) throws Exception {
+        boolean useVirtualThread = ((args.length > 0) && args[0].equals("Virtual"));
+        Thread thread;
+        System.out.println("Starting debuggee " + (useVirtualThread ? "virtual" : "platform") + " thread.");
+        if (useVirtualThread) {
+            thread = Thread.ofVirtual().name(THREAD_NAME).start(() -> { });
+        } else {
+            thread = Thread.ofPlatform().name(THREAD_NAME).start(() -> { });
+        }
+        thread.join();
+        done();
+    }
+}
+
+public class SuspendAfterDeath extends TestScaffold {
+    private volatile ThreadReference thread;
+    private volatile boolean breakpointReached = false;
+
+    SuspendAfterDeath(String args[]) {
+        super(args);
+    }
+
+    public static void main(String[] args) throws Exception {
+        new SuspendAfterDeath(args).startTests();
+    }
+
+    @Override
+    public void threadDied(ThreadDeathEvent event) {
+        ThreadReference eventThread = event.thread();
+        if (eventThread.name().equals(SuspendAfterDeathTarg.THREAD_NAME)) {
+            System.out.println("Target thread died, thread=" + eventThread);
+            thread = eventThread;
+        }
+    }
+
+    @Override
+    public void breakpointReached(BreakpointEvent event) {
+        ThreadReference eventThread = event.thread();
+        System.out.println("Breakpoint, thread=" + eventThread);
+        if (thread == null) {
+            failure("FAILED: got Breakpoint event before ThreadDeath event.");
+        }
+        breakpointReached = true;
+        /* Suspend the thread. This is being done after the thread has exited. */
+        thread.suspend();
+    }
+
+    @Override
+    public void connect(String args[]) {
+        String mainWrapper = System.getProperty("main.wrapper");
+        if ("Virtual".equals(mainWrapper)) {
+            List argList = new ArrayList(Arrays.asList(args));
+            argList.add("Virtual");
+            args = (String[]) argList.toArray(args);
```

Comment:
The raw type and casting caught my addition here. Here's something more succulent if you'd like:
```
args = Stream.concat(Stream.of(args), Stream.of("Virtual"))
                .toArray(String[]::new);
```

Reply:
That came from TestScaffold.startUp():
```
    protected void startUp(String targetName) {
        List argList = new ArrayList(Arrays.asList(args));
        argList.add(targetName);
        println("run args: " + argList);
        connect((String[]) argList.toArray(args));
        waitForVMStart();
    }
```
TBH I simply find myself staring at Streams code like that for way too long before I figure out what it is trying to do. If you use/read code like that a lot, it's probably not an issue, but I don't, and prefer something more straight forward, even if it is not nearly as elegant. Someone should add `String[] Arrays.appendToArray(String[] arr, String s)`, or even `String[] Arrays.concat(String[] s1, String[] s2)` would do.

Reply:
Whatever you are comfortable with but I think we should add the raw type and avoid the cast.

Reply:
fixed